### PR TITLE
Add documentation for HA cluster support features

### PR DESCRIPTION
**Pull request by barthisrael**
This PR adds documentation for all the features that have been as part
of the work of making it easier to integrate Barman with HA clusters.

We perform the following changes to the web docs / `man`:

* Add `cluster` configuration option to `man 5`
* Add `model` configuration option to `man 5`
* Include the `Model` scope to all options that configuration models support
* Add information about configuration models under the `Integration with cluster
  management systems` section of the web docs
* Add information about configuration models and examples under the `Configuration`
  section of the web docs
* Add documentation about the `barman config-switch` new CLI command in the
  web docs

Some changes to the documentation related with `wal_conninfo` and
`wal_streaming_conninfo` have already been applied through #868, and are only
pending to be merged.

Besides that we applied some standardization on all configuration options in `man`
(`doc/barman.5.d/50-*` files):

* Have an empty new line at the end -- some of them were missing this;
* Have at most 80 characters per line -- some of them were not following
  this rule;
* Have a `Scope: ` line near the end, after the description of the
  configuration option: it specifies which is the scope of the config
  option, i.e., if it can be applied to one or more scopes among
  `Global`, `Server` and `Model`. This information used to be present
  in the `man 5` files already, however they were part of the descritpion.
  Now we have some emphasis with a separate paragraph which follows the
  same standard over all files.

Also, there are more docs to come, once we finish other work related with HA cluster
features which is still pending.

References: BAR-136.

### CODE REVIEW

File: doc/manual/50-feature-details.en.md [hunk 1]
```diff
@@ -884,6 +884,59 @@ The `replication-status` command allows
...
+for example.
+
+As an example, let's say you have a PostgreSQL cluster with the following nodes:
+
+* `pg-node-1`: primary
+* `pg-node-2`: standby
+* `pg-node-3`: standby
+
+Assume you are backing up from the primary node, and have a configuration which
+include the following options:
```
Comment (martinmarques) [inline doc/manual/50-feature-details.en.md:902]:
```suggestion
includes the following options:
```

File: doc/manual/50-feature-details.en.md [hunk 2]
```diff
@@ -884,6 +884,59 @@ The `replication-status` command allows
...
+```
+
+Which could be applied upon a failover from `pg-node-1` to `pg-node-2` with
+the following command, so you start backing up from the new primary node:
+
+```bash
+barman config-switch my-barman-server my-barman-server:backup-from-pg-node-2
+```
+
+That will basically override the cluster configuration options with the values
```
Comment (martinmarques) [inline doc/manual/50-feature-details.en.md:929]:
```suggestion
That will override the cluster configuration options with the values
```

**Review by martinmarques — CHANGES_REQUESTED** [review]

File: doc/barman.5.d/50-cluster.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+cluster
+:   Name of the Barman cluster which a Barman server or model relates to. Used
```
Comment (mikewallace1979) [inline doc/barman.5.d/50-cluster.md:2]:
```suggestion
:   Name of the Barman cluster associated with a Barman server or model. Used
```

File: doc/manual/17-configuration.en.md [hunk 4]
```diff
@@ -6,14 +6,19 @@ There are two types of configuration files in Barman:
 
 - **global/general configuration**
 - **server configuration**
+- **model configuration**
```
Comment (mikewallace1979) [inline doc/manual/17-configuration.en.md:9]: [R5](https://github.com/EnterpriseDB/barman/pull/880/commits/a18f2bda214d20bcb14a0b703968c9d4e512cd29#diff-cb88ce549bfbef5a801899de30b88c00cb5f5544fbad2d1841b20cb24a0f20dcR5) should probably be updated to `There are three types of configuration files in Barman:`

File: doc/manual/17-configuration.en.md [hunk 5]
```diff
@@ -6,14 +6,19 @@ There are two types of configuration files in Barman:
...
 
 The main configuration file (set to `/etc/barman.conf` by default) contains general options such as main directory, system user, log file, and so on.
 
 Server configuration files, one for each server to be backed up by Barman, are located in the `/etc/barman.d` directory and must have a `.conf` suffix.
 
+Similarly, model configuration files, one for each model that can be applied to a server which is part of a cluster, are located in the `/etc/barman.d` directory and must have a `.conf` suffix.
```
Comment (mikewallace1979) [inline doc/manual/17-configuration.en.md:15]: This was a little hard to parse with the extra comma separated clause - I was thinking it could be reduced to:

```suggestion
Similarly, model configuration files are located in the `/etc/barman.d` directory and must have a `.conf` suffix.
```

I was thinking of adding the `one for each model that can be applied to a server which is part of a cluster` to the end, but I wonder whether one-model-per-file is something we should recommend at all. It might be better to group multiple models for a server (or cluster) in a single file (or maybe not...?).

File: doc/manual/21-preliminary_steps.en.md [hunk 6]
```diff
@@ -64,7 +64,7 @@ GRANT pg_checkpoint TO barman;
 > ["The Password File" section in the PostgreSQL Documentation][pgpass].
 
 This connection is required by Barman in order to coordinate its
-activities with the server, as well as for monitoring purposes.
+activities with the server, as well as for monitoring purposes.`
```
Comment (mikewallace1979) [inline doc/manual/21-preliminary_steps.en.md:67]: Is this backtick supposed to be here?

File: doc/manual/17-configuration.en.md [hunk 7]
```diff
@@ -6,14 +6,19 @@ There are two types of configuration files in Barman:
...
+
+> *NOTE*: models define a set of configuration overrides which can be applied on top of the configuration of Barman servers that are part of the same cluster as the model, through the `barman config-switch` command.
```
Comment (mikewallace1979) [inline doc/manual/17-configuration.en.md:17]: Could we link to the `barman config-switch` section here?

File: doc/manual/42-server-commands.en.md [hunk 8]
```diff
@@ -67,6 +67,40 @@ barman check <server_name>
 > and monitoring infrastructure. The `--nagios` option allows you
 > to easily create a plugin for Nagios/Icinga.
 
+## `config-switch`
+
+The `config-switch` command is used to apply a set of configuration overrides
+defined through a model to a Barman server. The final configuration of the Barman
+server is composed by the configuration of the server plus the overrides applied
```
Comment (mikewallace1979) [inline doc/manual/42-server-commands.en.md:74]:
```suggestion
server is composed of the configuration of the server plus the overrides applied
```

File: doc/manual/42-server-commands.en.md [hunk 9]
```diff
@@ -67,6 +67,40 @@ barman check <server_name>
...
+by the selected model. Models are specially useful for clustered environments,
```
Comment (mikewallace1979) [inline doc/manual/42-server-commands.en.md:75]:
```suggestion
by the selected model. Models are particularly useful for clustered environments,
```

File: doc/manual/50-feature-details.en.md [hunk 10]
```diff
@@ -884,6 +884,59 @@ The `replication-status` command allows
 you to get information about any streaming client attached to the
 managed server, in particular hot standby servers and WAL streamers.
 
+### Configuration Models
+
+Configuration models define a set of overrides for configuration options. These
+overrides can be applied to Barman servers which are part of the same cluster as
+the config models. They can be handy to handle clustered environments, so you
```
Comment (mikewallace1979) [inline doc/manual/50-feature-details.en.md:891]:
```suggestion
the config models. They can be useful when handling clustered environments, so you
```

File: doc/manual/50-feature-details.en.md [hunk 11]
```diff
@@ -884,6 +884,59 @@ The `replication-status` command allows
...
+; other options...
+```
+
+You could, for example, have a configuration model for that cluster as follows:
+
+```ini
+[my-barman-server:backup-from-pg-node-2]
+cluster = my-cluster
+model = true
+conninfo = host=pg-node-2 user=barman
```
Comment (mikewallace1979) [inline doc/manual/50-feature-details.en.md:918]: It might be worth keeping the conninfo the same apart from the host, just in case a user misinterprets that the `database=postgres` is omitted from conninfo here because it's present in the conninfo in the server configuration.

```suggestion
conninfo = host=pg-node-2 user=barman database=postgres
```

**Review by mikewallace1979 — COMMENTED** [review]
Looks good - I had a few minor wording suggestions but nothing major.

File: doc/manual/17-configuration.en.md (hunk 5 above)
Comment (barthisrael) [reply doc/manual/17-configuration.en.md:15]: I wonder if "one model per file" wouldn't end up bloating the directory 😆 I guess we can just keep the way you suggested above, and add nothing else. What do you think?

File: doc/manual/21-preliminary_steps.en.md (hunk 6 above)
Comment (barthisrael) [reply doc/manual/21-preliminary_steps.en.md:67]: Typo, good catch.

File: doc/manual/17-configuration.en.md (hunk 7 above)
Comment (barthisrael) [reply doc/manual/17-configuration.en.md:17]: I think that makes sense, I'll take a look at adding a ref.

File: doc/manual/50-feature-details.en.md (hunk 11 above)
Comment (barthisrael) [reply doc/manual/50-feature-details.en.md:918]: Makes sense!

File: doc/manual/17-configuration.en.md (hunk 7 above)
Comment (barthisrael) [reply doc/manual/17-configuration.en.md]: Applied.

File: doc/manual/21-preliminary_steps.en.md (hunk 6 above)
Comment (barthisrael) [reply doc/manual/21-preliminary_steps.en.md]: Applied.

File: doc/manual/17-configuration.en.md (hunk 4 above)
Comment (barthisrael) [reply doc/manual/17-configuration.en.md:9]: Applied.

File: doc/manual/17-configuration.en.md (hunk 5 above)
Comment (mikewallace1979) [reply doc/manual/17-configuration.en.md]: Yeah that's fine I think.

**Comment by barthisrael** [comment]
We will merge this once #868 and #873 get merged because this PR includes documentation for those PRs.
We decided to merge this PR and follow-up with documenting future work related with models through the specific PRs that are coming.

**Review by mikewallace1979 — APPROVED** [review]

**Review by gcalacoci — APPROVED** [review]
LGTM

**Comment by barthisrael** [comment]
I'm bypassing protections and merging this anyway (Martin is currently on PTO and his thoughts have been applied to the PR).
